### PR TITLE
ARROW-9350: [C++][CI] Nightly valgrind job failures

### PR DESCRIPTION
**Pull request by kszucs**
The builds have been failing since https://github.com/apache/arrow/commit/d0f3b5f3c74f67c7ca941c98bd60148e9a9e94f0

### CODE REVIEW

File: cpp/src/arrow/array/array_base.h [hunk 1]
```diff
@@ -176,7 +176,7 @@ class ARROW_EXPORT Array {
   Status ValidateFull() const;
 
  protected:
-  Array() : null_bitmap_data_(NULLPTR) {}
+  Array() : data_(NULLPTR), null_bitmap_data_(NULLPTR) {}
```
Comment (kszucs) [inline cpp/src/arrow/array/array_base.h:179]: There are multiple `Conditional jump or move depends on uninitialised value(s)` from `Array.IsNull` and `Array.IsValid` but it doesn't seem to solve it.

Comment (pitrou) [reply cpp/src/arrow/array/array_base.h:179]: Yes, this change is not useful as `shared_ptr` is always initialized.

Comment (kszucs) [reply cpp/src/arrow/array/array_base.h]: I initialized as ArrayData because IsNull references `data_->offset`

Comment (pitrou) [reply cpp/src/arrow/array/array_base.h]: I don't think that would solve anything. As I said, `shared_ptr` is always initialized (to nullptr).

**Comment by pitrou** [comment]
Should probably close this in favor of #7679

**Comment by kszucs** [comment]
Closing.

**Pull request closed without merging**